cloudflare/resource_cloudflare_access_mutual_tls_certificate.go

```diff
@@ -157,18 +159,45 @@ func resourceCloudflareAccessMutualTLSCertificateDelete(d *schema.ResourceData,
 		return err
 	}
 
+	// To actually delete the certificate, it cannot have any hostnames associated
+	// with it so here we perform an update (to remove them) before we continue on
+	// with wiping the certificate itself.
+	deletedCertificate := cloudflare.AccessMutualTLSCertificate{
+		ID:                  d.Id(),
+		Name:                d.Get("name").(string),
+		AssociatedHostnames: []string{},
+	}
+
 	if identifier.Type == AccountType {
-		err = client.DeleteAccessMutualTLSCertificate(context.Background(), identifier.Value, certID)
+		_, err = client.UpdateAccessMutualTLSCertificate(context.Background(), identifier.Value, d.Id(), deletedCertificate)
 	} else {
-		err = client.DeleteZoneAccessMutualTLSCertificate(context.Background(), identifier.Value, certID)
+		_, err = client.UpdateZoneAccessMutualTLSCertificate(context.Background(), identifier.Value, d.Id(), deletedCertificate)
 	}
+
 	if err != nil {
-		return fmt.Errorf("error deleting Access Mutual TLS Certificate for %s %q: %s", identifier.Type, identifier.Value, err)
+		return fmt.Errorf("error updating Access Mutual TLS Certificate for %s %q: %s", identifier.Type, identifier.Value, err)
 	}
 
-	d.SetId("")
+	return resource.Retry(d.Timeout(schema.TimeoutDelete), func() *resource.RetryError {
```

Comment:
We implement a `Retry` here in the event that the service has lag removing the associations in order to check again after a small wait.